File: Olympiad/COCI/official/2009/contest5_solutions/jagoda.pas
program jagoda;

const MAXN = 100000;

var
   box, cup              : array[0..MAXN] of longint;
   N, M, K, A, B, ans, i : longint;

begin
   read(N, M);
   K := trunc(sqrt(N));
   for i:=1 to M do begin
      read(B, A);
      dec(A);
      B := B + A;

      ans := 0;
      while (A < B) and (A mod K <> 0) do begin
         inc(box[A]);
         ans := ans + box[A];
         inc(A);
      end;

      if B <> N then begin
         while (A < B) and (B mod K <> 0) do begin
            dec(B);
            inc(box[B]);
            ans := ans + box[B];
         end;
      end;

      while A < B do begin
         inc(cup[A div K]);
         ans := ans + cup[A div K];
         A := A + K;
      end;
      writeln(ans);
   end;
end.

File: Olympiad/COCI/official/2009/contest5_solutions/jagoda.cpp
#include <cmath>
#include <iostream>
using namespace std;

#define MAXN 100000

int box[MAXN], cup[MAXN];

int main() {
   cin.sync_with_stdio(0);

   int N, M;
   cin >> N >> M;

   int K = (int)sqrt(N);
   while ( M-- > 0 ) {
      int B, A;
      cin >> B >> A;
      --A; B += A;

      int ans = 0;
      while ( A < B && A%K != 0 ) ans += ++box[A++];
      if ( B != N ) {
         while ( A < B && B%K != 0 ) ans += ++box[--B];
      }
      while ( A < B ) ans += ++cup[A/K], A += K;
      cout << ans << endl;
   }

   return 0;
}
